File: shell/gnome/src/donutGeometry.js
const HOLE_RATIO = 0.618;
const GAP_RATIO = 2 / 104;
const CORNER_RATIO = 0.2;
const MIN_WIDTH_RATIO = 3 / 104;
const FULL_TURN = 2 * Math.PI;
const QUARTER_TURN = Math.PI / 2;
const ANGLE_EPSILON = 1e-9;
export const START_ANGLE = -QUARTER_TURN;

export function donutGeometry(size) {
    const outer = size / 2;
    const thickness = outer * (1 - HOLE_RATIO);
    return {
        center: outer,
        outer,
        inner: outer - thickness,
        thickness,
        gap: GAP_RATIO * size,
        corner: CORNER_RATIO * thickness,
    };
}

const UNIT = donutGeometry(1);
export const MIN_FRACTION = (MIN_WIDTH_RATIO + UNIT.gap) / UNIT.inner / FULL_TURN;

function minimumFraction(count) {
    return count > 1 ? Math.min(1 / count, MIN_FRACTION) : 0;
}

function shares(values, raised, minimum) {
    const free = values.map((value, index) => (raised.has(index) ? 0 : value));
    const freeTotal = free.reduce((sum, value) => sum + value, 0);
    const share = 1 - raised.size * minimum;
    return values.map((_value, index) => (raised.has(index) ? minimum : (free[index] / freeTotal) * share));
}

function spread(values, minimum) {
    const raised = new Set();
    const tooSmall = fractions => fractions.findIndex((fraction, index) => !raised.has(index) && fraction < minimum);
    let fractions = shares(values, raised, minimum);
    for (let index = tooSmall(fractions); index >= 0; index = tooSmall(fractions)) {
        raised.add(index);
        fractions = shares(values, raised, minimum);
    }
    return fractions;
}

export function visibleFractions(values) {
    const total = values.reduce((sum, value) => sum + value, 0);
    if (total <= 0) return values.map(() => 0);
    return spread(values, minimumFraction(values.length));
}

export function donutSegments(fractions, reveal = 1) {
    if (reveal <= 0) return [];
    if (fractions.length === 1) {
        const end = START_ANGLE + Math.min(reveal, fractions[0]) * FULL_TURN;
        return end > START_ANGLE ? [{ index: 0, start: START_ANGLE, end, gap: false }] : [];
    }
    const limit = START_ANGLE + reveal * FULL_TURN;
    const segments = [];
    let angle = START_ANGLE;
    fractions.forEach((fraction, index) => {
        const start = angle;
        angle += fraction * FULL_TURN;
        const end = Math.min(angle, limit);
        if (end > start) segments.push({ index, start, end, gap: true });
    });
    return segments;
}

function cornerRadius(geometry, halfSweep, halfGap) {
    const sin = Math.sin(Math.min(halfSweep, QUARTER_TURN));
    const innerRoom = (geometry.inner * sin - halfGap) / (1 - sin);
    const outerRoom = (geometry.outer * sin - halfGap) / (1 + sin);
    const room = Math.min(innerRoom, outerRoom);
    return room < 0 ? null : Math.min(geometry.corner, room);
}

function polar(geometry, radius, angle) {
    return [geometry.center + radius * Math.cos(angle), geometry.center + radius * Math.sin(angle)];
}

function ringPath(geometry) {
    const center = [geometry.center, geometry.center];
    return [
        { center, radius: geometry.outer, from: 0, to: FULL_TURN },
        { center, radius: geometry.inner, from: FULL_TURN, to: 0, negative: true, newSubPath: true },
    ];
}

function sectorArcs(geometry, { start, end }, corner, outerOffset, innerOffset) {
    const center = [geometry.center, geometry.center];
    const outerCorner = angle => polar(geometry, geometry.outer - corner, angle);
    const innerCorner = angle => polar(geometry, geometry.inner + corner, angle);
    const fillet = (at, from, to) => ({ center: at, radius: corner, from, to });
    return [
        fillet(outerCorner(start + outerOffset), start - QUARTER_TURN, start + outerOffset),
        {
            center,
            radius: geometry.outer,
            from: start + outerOffset,
            to: Math.max(end - outerOffset, start + outerOffset),
        },
        fillet(outerCorner(end - outerOffset), end - outerOffset, end + QUARTER_TURN),
        fillet(innerCorner(end - innerOffset), end + QUARTER_TURN, end - innerOffset + Math.PI),
        {
            center,
            radius: geometry.inner,
            from: end - innerOffset,
            to: Math.min(start + innerOffset, end - innerOffset),
            negative: true,
        },
        fillet(innerCorner(start + innerOffset), start + innerOffset + Math.PI, start + 3 * QUARTER_TURN),
    ];
}

export function sectorPath(geometry, segment) {
    if (!segment.gap && segment.end - segment.start >= FULL_TURN - ANGLE_EPSILON) return ringPath(geometry);
    const halfGap = segment.gap ? geometry.gap / 2 : 0;
    const corner = cornerRadius(geometry, (segment.end - segment.start) / 2, halfGap);
    if (corner === null) return null;
    const outerOffset = Math.asin((halfGap + corner) / (geometry.outer - corner));
    const innerOffset = Math.asin((halfGap + corner) / (geometry.inner + corner));
    return sectorArcs(geometry, segment, corner, outerOffset, innerOffset);
}

File: shell/macos/Sources/HeadroomKit/Presentation/DonutGeometry.swift
import Foundation

public struct DonutSegment: Sendable, Hashable {
    public let index: Int
    public let start: Double
    public let end: Double
    public let gap: Bool
}

public struct DonutGeometry: Sendable, Hashable {
    public static let holeRatio = 0.618
    public static let startAngle = -Double.pi / 2
    static let gapRatio = 2.0 / 104
    static let cornerRatio = 0.15
    static let minWidthRatio = 3.0 / 104
    static let fullTurn = 2 * Double.pi
    static let quarterTurn = Double.pi / 2
    static let angleEpsilon = 1e-9

    public let center: Double
    public let outer: Double
    public let inner: Double
    public let gap: Double
    public let corner: Double

    public init(size: Double) {
        outer = size / 2
        center = outer
        let thickness = outer * (1 - Self.holeRatio)
        inner = outer - thickness
        gap = Self.gapRatio * size
        corner = Self.cornerRatio * thickness
    }

    static var minimumFraction: Double {
        let unit = DonutGeometry(size: 1)
        return (minWidthRatio + unit.gap) / unit.inner / fullTurn
    }

    public static func visibleFractions(_ values: [Int64]) -> [Double] {
        let doubles = values.map { Double(max(0, $0)) }
        let total = doubles.reduce(0, +)
        guard total > 0 else { return values.map { _ in 0 } }
        let minimum = values.count > 1 ? min(1 / Double(values.count), minimumFraction) : 0
        return spread(doubles, minimum: minimum)
    }

    private static func spread(_ values: [Double], minimum: Double) -> [Double] {
        var raised = Set<Int>()
        var fractions = shares(values, raised: raised, minimum: minimum)
        while let index = fractions.indices.first(where: { !raised.contains($0) && fractions[$0] < minimum }) {
            raised.insert(index)
            fractions = shares(values, raised: raised, minimum: minimum)
        }
        return fractions
    }

    private static func shares(_ values: [Double], raised: Set<Int>, minimum: Double) -> [Double] {
        let freeTotal = values.indices.filter { !raised.contains($0) }.map { values[$0] }.reduce(0, +)
        let share = 1 - Double(raised.count) * minimum
        return values.indices.map { index in
            if raised.contains(index) { return minimum }
            return freeTotal > 0 ? values[index] / freeTotal * share : 0
        }
    }

    public static func segments(_ fractions: [Double], reveal: Double = 1) -> [DonutSegment] {
        guard reveal > 0 else { return [] }
        if fractions.count == 1 {
            let end = startAngle + min(reveal, fractions[0]) * fullTurn
            return end > startAngle ? [DonutSegment(index: 0, start: startAngle, end: end, gap: false)] : []
        }
        let limit = startAngle + reveal * fullTurn
        var angle = startAngle
        var result: [DonutSegment] = []
        for (index, fraction) in fractions.enumerated() {
            let start = angle
            angle += fraction * fullTurn
            let end = min(angle, limit)
            if end > start { result.append(DonutSegment(index: index, start: start, end: end, gap: true)) }
        }
        return result
    }
}
